src/Razor/Microsoft.AspNetCore.Razor.Language/test/Legacy/RazorDirectivesTest.cs

```diff
@@ -907,6 +907,79 @@ public void Directives_ReservedWordInsideCodeBlock()
             ParseDocumentTest("@{ class }");
         }
 
+        [Fact]
+        public void DirectiveDescriptor_UnderstandsGenericConstraintsToken()
+        {
+            // Arrange
+            var descriptor = DirectiveDescriptor.CreateDirective(
+                "custom",
+                DirectiveKind.SingleLine,
+                b => {
+                    b.AddMemberToken();
+                    b.AddOptionalGenericTypeConstraintToken("name", "description");
+                });
+
+            // Act & Assert
+            ParseDocumentTest(@"
+@custom TSomething where TSomething : class
+",
+                new[] { descriptor });
+        }
+
+        [Fact]
+        public void DirectiveDescriptor_GenericConstraintTokenWorksWhenAtEndOfFile()
+        {
+            // Arrange
+            var descriptor = DirectiveDescriptor.CreateDirective(
+                "custom",
+                DirectiveKind.SingleLine,
+                b => {
+                    b.AddMemberToken();
+                    b.AddOptionalGenericTypeConstraintToken("name", "description");
+                });
+
+            // Act & Assert
+            ParseDocumentTest(@"
+@custom TSomething where TSomething : class",
+                directives: new[] { descriptor });
+        }
+
+        [Fact]
+        public void DirectiveDescriptor_GenericConstraintTokenProducesErrorWhenFirstTokenIsNotWhereKeyword()
+        {
+            // Arrange
+            var descriptor = DirectiveDescriptor.CreateDirective(
+                "custom",
+                DirectiveKind.SingleLine,
+                b => {
+                    b.AddMemberToken();
+                    b.AddOptionalGenericTypeConstraintToken("name", "description");
+                });
+
+            // Act & Assert
+            ParseDocumentTest(@"
+@custom TSomething maybe TSomething : class",
+                directives: new[] { descriptor });
+        }
+
+        [Fact]
+        public void DirectiveDescriptor_GenericConstraintTokenProducesErrorWhenConstraintIdentifierIsDifferentFromPrecedingMember()
```

Comment:
nit: do we really need 13 word test names haha?  perhaps GenericConstraintErrorsWhenIdentifierDifferent is good enough?

Reply:
done